public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx

```diff
@@ -0,0 +1,118 @@
+import { css } from '@emotion/css';
+import React from 'react';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Button, Card, useStyles2 } from '@grafana/ui';
+import { RawQuery } from 'app/plugins/datasource/prometheus/querybuilder/shared/RawQuery';
+
+import promqlGrammar from '../promql';
+
+import { promQueryModeller } from './PromQueryModeller';
+import { PromQueryPattern } from './types';
+
+type Props = {
+  pattern: PromQueryPattern;
+  hasNewQueryOption: boolean;
+  hasPreviousQuery: boolean;
+  selectedPatternName: string | null;
+  setSelectedPatternName: (name: string | null) => void;
+  onPatternSelect: (pattern: PromQueryPattern, selectAsNewQuery?: boolean) => void;
+};
+
+export const QueryPattern = (props: Props) => {
+  const { pattern, onPatternSelect, hasNewQueryOption, hasPreviousQuery, selectedPatternName, setSelectedPatternName } =
+    props;
+
+  const styles = useStyles2(getStyles);
+  const lang = { grammar: promqlGrammar, name: 'promql' };
+
+  return (
+    <Card className={styles.card}>
+      <Card.Heading>{pattern.name}</Card.Heading>
+      <div className={styles.rawQueryContainer}>
+        <RawQuery
+          aria-label={`${pattern.name} raw query`}
+          query={promQueryModeller.renderQuery({
+            labels: [],
+            operations: pattern.operations,
+            binaryQueries: pattern.binaryQueries,
+          })}
+          lang={lang}
+          className={styles.rawQuery}
+        />
+      </div>
+      <Card.Actions>
+        {selectedPatternName !== pattern.name ? (
+          <Button
+            size="sm"
+            aria-label="use this query button"
+            onClick={() => {
+              if (hasPreviousQuery) {
```

Comment:
This hasPreviousQuery only seems to work when they've accepted a previous query pattern, would we want to show this whenever there's any text in the query editor (basically always)?

Reply:
Good point! It does only work when they've accepted a previous query pattern and thinking about it, it doesn't actually replace the query, it only applies the chosen query pattern to what is already there! Like, it you choose sum then rate, it does that, then choose histogram to "replace" the query, you get a histogram of the sum and rate! You're right, if there is a query already entered, we should use this functionality, whether it be a metric or a previous query pattern. Also, I think we should change the wording to from "replace this query" to "apply this pattern."

Reply:
I made the change, this now asks a user to apply the query pattern if there is a previous metric, label, operation or binary query in the editor whereas previously it only checked operations.